File: .kin-snapshot-tmp/tests/adapter-fixtures/java/EdgeCases.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Firelock, LLC

// Fixture: Java edge cases
// Tests: interfaces, generics, annotations, enums with methods, inner classes,
// abstract classes, default methods, static methods, lambdas, records

package com.example.shapes;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;

/** Interface with default and static methods. */
public interface Shape {
    double area();
    double perimeter();

    default String describe() {
        return String.format("%s(area=%.2f)", getClass().getSimpleName(), area());
    }

    static Shape larger(Shape a, Shape b) {
        return a.area() >= b.area() ? a : b;
    }
}

/** Abstract base with template method. */
abstract class AbstractShape implements Shape {
    private final String color;

    protected AbstractShape(String color) {
        this.color = color;
    }

    public String getColor() {
        return color;
    }

    @Override
    public String toString() {
        return describe() + " [" + color + "]";
    }
}

/** Concrete class with generics in constructor. */
class Circle extends AbstractShape {
    private final double radius;

    public Circle(double radius) {
        super("red");
        this.radius = radius;
    }

    @Override
    public double area() {
        return Math.PI * radius * radius;
    }

    @Override
    public double perimeter() {
        return 2 * Math.PI * radius;
    }

    /** Factory method. */
    public static Circle unit() {
        return new Circle(1.0);
    }
}

/** Enum with fields and methods. */
enum Color {
    RED("#FF0000"),
    GREEN("#00FF00"),
    BLUE("#0000FF");

    private final String hex;

    Color(String hex) {
        this.hex = hex;
    }

    public String getHex() {
        return hex;
    }

    public boolean isWarm() {
        return this == RED;
    }
}

/** Generic container. */
class Container<T extends Comparable<T>> {
    private final List<T> items;

    public Container(List<T> items) {
        this.items = items;
    }

    public Optional<T> max() {
        return items.stream().max(Comparable::compareTo);
    }

    /** Inner class. */
    class Iterator {
        private int index = 0;

        public boolean hasNext() {
            return index < items.size();
        }

        public T next() {
            return items.get(index++);
        }
    }
}

/** Annotation definition. */
@interface Validate {
    String message() default "validation failed";
    int priority() default 0;
}

/** Class using custom annotation. */
class Processor {
    @Validate(message = "input required", priority = 1)
    public String process(String input) {
        return input.toUpperCase();
    }

    /** Static final constant. */
    public static final int MAX_RETRIES = 3;

    /** Lambda usage. */
    public Function<String, String> getTransformer() {
        return s -> s.trim().toLowerCase();
    }
}

File: .kin-snapshot-tmp/tests/adapter-fixtures/rust/edge_cases.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Firelock, LLC

// Fixture: Rust edge cases
// Tests: trait impls, derive macros, generics, lifetimes, enum variants,
// closures, async, nested modules, where clauses, const generics

use std::collections::HashMap;
use std::fmt;

/// A generic container with a trait bound.
pub struct Container<T: Clone + fmt::Display> {
    items: Vec<T>,
    label: String,
}

impl<T: Clone + fmt::Display> Container<T> {
    pub fn new(label: &str) -> Self {
        Self {
            items: Vec::new(),
            label: label.to_string(),
        }
    }

    pub fn push(&mut self, item: T) {
        self.items.push(item);
    }

    pub fn first(&self) -> Option<&T> {
        self.items.first()
    }
}

/// Display impl for Container.
impl<T: Clone + fmt::Display> fmt::Display for Container<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Container({}): {} items", self.label, self.items.len())
    }
}

/// An enum with data variants.
pub enum Shape {
    Circle { radius: f64 },
    Rectangle { width: f64, height: f64 },
    Triangle(f64, f64, f64),
}

impl Shape {
    pub fn area(&self) -> f64 {
        match self {
            Shape::Circle { radius } => std::f64::consts::PI * radius * radius,
            Shape::Rectangle { width, height } => width * height,
            Shape::Triangle(a, b, c) => {
                let s = (a + b + c) / 2.0;
                (s * (s - a) * (s - b) * (s - c)).sqrt()
            }
        }
    }
}

/// Trait with associated type and default method.
pub trait Processor {
    type Output;
    type Error;

    fn process(&self, input: &str) -> Result<Self::Output, Self::Error>;

    fn name(&self) -> &str {
        "unnamed"
    }
}

/// Function with lifetime parameters.
pub fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() > y.len() { x } else { y }
}

/// Function with where clause.
pub fn serialize_map<K, V>(map: &HashMap<K, V>) -> String
where
    K: fmt::Display,
    V: fmt::Display,
{
    map.iter()
        .map(|(k, v)| format!("{}: {}", k, v))
        .collect::<Vec<_>>()
        .join(", ")
}

/// Async function.
pub async fn fetch_data(url: &str) -> Result<String, Box<dyn std::error::Error>> {
    Ok(format!("data from {}", url))
}

/// Constant with type annotation.
pub const MAX_RETRIES: u32 = 3;

/// Static mutable (unsafe).
pub static COUNTER: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(0);

/// Nested module.
pub mod utils {
    pub fn helper() -> bool {
        true
    }
}

/// Type alias with generics.
pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn container_works() {
        let mut c = Container::new("test");
        c.push(42);
        assert_eq!(c.first(), Some(&42));
    }

    #[test]
    fn shape_area() {
        let circle = Shape::Circle { radius: 1.0 };
        assert!((circle.area() - std::f64::consts::PI).abs() < 1e-10);
    }
}
